Improve autoconfiguration capabilities for adding, delaying or dropping spans

JonasKunz opened the pull request:
The goal of this PR is to start a discussion around enhancing the ability to modify spans before export.

The existing mechanism to do this are the following: 
 * SpanProcessor for modifying spans on start
 * wrapping `SpanExporters` for modifying spans before they are exported (= after they have ended)

These options only allow to modify the contents of spans, but they do not allow to efficiently perform the following changes:
 * Filtering spans, so that they are not exported
    * See for example [this slack discussion](https://cloud-native.slack.com/archives/C014L2KCTE3/p1671486545789789)
 * Delaying spans
   * E.g. in order to add profiling statistics to spans as attributes. Those statistics are likely not immediately available on span end.
 * Adding artificial spans
   * Example use-case: Generating spans from profiling data. Could be done via a normal `Tracer`, but this would technically break the `SpanProcessor` contracts because `onStart` / `onEnd` will not be invoked synchronously on the thread to which the span "belongs".

In theory these actions can already be performed by wrapping `SpanExporters`. This is however not a good idea, because the addition, delay or dropping would happen *after* the batching performed by `BatchSpanProcessor`.
As a result, it essentially breaks batching, making it not really suitable for production.

A batching-friendly alternative would be to wrap the `BatchSpanProcessor` instead of the `SpanExporters`:

```
BatchSpanProcessor exportProcessor = BatchSpanProcessor.builder(myExporter).build();

SdkTracerProvider.builder()
  ...
  .addSpanProcessor(new MyWrapper(exportProcessor))
```

This approach already works nicely for manually configured SDK instances.
Unfortunately there is no corresponding hook available in the SDK autoconfiguration mechanism.
The code changes in this PR adds a corresponding hook, which could be made experimental for a start.

~Note that this PR is intended as just a starting point for a discussion rather than a ready-to-merge suggestion. I'm more than happy to discuss totally different alternatives.~

~If we decide to go the route suggested in this PR, I'll add test cases which I have omitted for a start.~

PR has been updated and enhanced with a test case.

trask commented on sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/AutoConfigurationCustomizer.java:
```diff
@@ -62,6 +63,21 @@ AutoConfigurationCustomizer addSpanExporterCustomizer(
+   *
+   * <p>Multiple calls will execute the customizers in order.
+   */
+  AutoConfigurationCustomizer addSpanExportProcessorCustomizer(
```
?
```suggestion
  AutoConfigurationCustomizer addBatchSpanProcessorCustomizer(
```

jackshirazi commented:
@JonasKunz please implement more completely

JonasKunz commented:
@open-telemetry/java-maintainers   I've completed the implementation, added a test-case and fixed the failing checks as requested in the last SIG-Meeting. Let me know if you need anything else from my side to move ahead with this PR.

jack-berg commented on sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/AutoConfigurationCustomizer.java:
```diff
@@ -62,6 +63,23 @@ AutoConfigurationCustomizer addSpanExporterCustomizer(
+   *
+   * <p>Multiple calls will execute the customizers in order.
+   */
+  default AutoConfigurationCustomizer addBatchSpanProcessorCustomizer(
```
```suggestion
  default AutoConfigurationCustomizer addSpanProcessorCustomizer(
```

trask replied:
will this customizer be applied to all span processors, or only to the batch span processor?

jack-berg replied:
As written in this PR currently it applies to the result of `SpanProcessor.composite`, which may be a batch processor or a composite processor depending on whether one or multiple exporters are configured.

I don't see why we would want to limit it to batch processor. Right now, there is no way for other processors to be automatically configured besides simple and batch processor, but with file configuration that will change. If we end up providing the ability to customize the components (notably, I don't think we should allow customizing components), then there will be other potential processors besides simple and batch.

JonasKunz replied:
Why did I go for customizing the result of `SpanProcessor.composite` instead of the individual span processors?
Consider the use-case from my initial comment:
> Delaying spans
E.g. in order to add profiling statistics to spans as attributes. Those statistics are likely not immediately available on span end.

The idea here was, that even if multiple span-exporting `SpanProcessors` are registered, the "delaying" done by the customizer and resulting buffering only needs to be done once per span.
If you instead wrap each individual `SpanProcessor`, you'll need to do the buffering for each one: E.g. if you have both a `SimpleSpanProcessor` and a `BatchSpanProcessor` registered, you would end up with double the memory consumed for buffering and doing the correlation with the data you are waiting for two times.

In production environments I would however expect that you only have a single `BatchSpanProcessor` and no `SimpleSpanProcessor`, so the proposal to wrap individual `SpanProcessors` is fine for me. I'll then just remove the following guarantee from the javadoc:
```
The provided function will be invoked at most once per customized SDK instance.
```

> Right now, there is no way for other processors to be automatically configured besides simple and batch processor, but with file configuration that will change.

I'm curious, do you have any link to an Issue/PR/something to read up on that feature?

I originally named the method `addSpanExportingProcessorCustomizer` to explicitly state that the customizer would only be invoked for `SpanProcessor`s responsible for invoking `SpanExporter`s. The reason is that I did not want to wrap other `SpanProcessors`, e.g. ones responsible for lifting baggage into span-attributes.

I guess with your proposal this still can be done by doing an `instanceof BatchSpanProcessor` check in the customizer, though this feels a little hacky.

jack-berg replied:
>> Right now, there is no way for other processors to be automatically configured besides simple and batch processor, but with file configuration that will change.

> I'm curious, do you have any link to an Issue/PR/something to read up on that feature?

Here are some resources:
- The original [OTEP](https://github.com/open-telemetry/oteps/blob/main/text/0225-configuration.md)
- The relevant portion of the [specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/file-configuration.md)
- The [opentelemetry-configuration](https://github.com/open-telemetry/opentelemetry-configuration) repository defining the file schema
- The Configuration SIG meets every other monday at 8am PT. Details [here](https://github.com/open-telemetry/community#specification-sigs).

> I guess with your proposal this still can be done by doing an instanceof BatchSpanProcessor check in the customizer, though this feels a little hacky.

This is the expectation right now with all the customizers (i.e. `addSamplerCustomizer`, `add{Signal}ExporterCustomizer`, `addPropagatorCustomizer`), so I think its the right decision.

Another feature of the customizers is they give you the ability to fully replace the component in addition to the ability to wrap it with a delegating class. I'm imagine you could use this with processor customizer for an additional use case: Suppose you don't want to use the batch processor at all, but you know that autoconfigured exporters always get paired with a batch processor. If we add an additional `getExporter()` method to `BatchSpanProcessor`, then you could use a span processor customizer to replace the batch processor with another, while retaining the autoconfigured exporter. Something like:
```
    autoConfiguration.addSpanProcessorCustomizer(
        (spanProcessor, configProperties) -> {
          if (spanProcessor instanceof BatchSpanProcessor) {
            SpanExporter exporter = ((BatchSpanProcessor) spanProcessor).getExporter();
            return new MyCustomProcessor(exporter);
          }
          return spanProcessor;
        });
```
Pretty slick, IMO.

jack-berg commented on sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/AutoConfiguredOpenTelemetrySdkBuilder.java:
```diff
@@ -71,6 +72,9 @@ public final class AutoConfiguredOpenTelemetrySdkBuilder implements AutoConfigur
       spanExporterCustomizer = (a, unused) -> a;
+
+  private BiFunction<? super SpanProcessor, ConfigProperties, ? extends SpanProcessor>
+      batchSpanProcessorCustomizer = (a, unused) -> a;
```
```suggestion
      spanProcessorCustomizer = (a, unused) -> a;
```

jack-berg commented on sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/TracerProviderConfiguration.java:
```diff
@@ -53,11 +55,17 @@ static void configureTracerProvider(
-    configureSpanProcessors(config, exportersByName, meterProvider, closeables)
-        .forEach(tracerProviderBuilder::addSpanProcessor);
+    List<SpanProcessor> processors =
+        configureExportingSpanProcessors(config, exportersByName, meterProvider, closeables);
```
Let's apply the spanProcessorCustomizer to each span processor individually, rather than to the composite processor.

jack-berg commented on sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/AutoConfigurationCustomizer.java:
```diff
@@ -62,6 +63,23 @@ AutoConfigurationCustomizer addSpanExporterCustomizer(
+   *
+   * <p>Multiple calls will execute the customizers in order.
+   */
+  default AutoConfigurationCustomizer addBatchSpanProcessorCustomizer(
```
For symmetry, let's go ahead and add `addLogRecordProcessorCustomizer` as well.
